<a href="https://colab.research.google.com/github/davidpedem/masterPLIA2324-traduccion-automatica-sesion-1-davidpedem/blob/main/Ejercicio_Comparativa_Traducciones.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Evaluación Comparativas Traducciones
#David Peña de Miguel

In [ ]:
!pip install evaluate -Uq

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 1.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 536.7/536.7 kB 19.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 12.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 15.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.3/38.3 MB 33.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ibis-framework 7.1.0 requires pyarrow<15,>=2, but you have pyarrow 15.0.0 which is incompatible.


In [ ]:
!pip install jiwer bert_score git+https://github.com/google-research/bleurt.git -Uq

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 3.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 28.8 MB/s eta 0:00:00


In [ ]:
original = "In a hole in the ground there lived a hobbit. Not a nasty, dirty, wet hole, filled with the ends of worms and an oozy smell, nor yet a dry, bare, sandy hole with nothing in it to sit down on or to eat: it was a hobbit-hole, and that means comfort. It had a perfectly round door like a porthole, painted green, with a shiny yellow brass knob in the exact middle. The door opened on to a tube-shaped hall like a tunnel: a very comfortable tunnel without smoke, with panelled walls, and floors tiled and carpeted, provided with polished chairs, and lots and lots of pegs for hats and coats - the hobbit was fond of visitors. The tunnel wound on and on, going fairly but not quite straight into the side of the hill - The Hill, as all the people for many miles round called it - and many little round doors opened out of it, first on one side and then on another. No going upstairs for the hobbit: bedrooms, bathrooms, cellars, pantries (lots of these), wardrobes (he had whole rooms devoted to clothes), kitchens, dining-rooms, all were on the same floor, and indeed on the same passage. The best rooms were all on the lefthand side (going in), for these were the only ones to have windows, deep-set round windows looking over his garden, and meadows beyond, sloping down to the river. This hobbit was a very well-to-do hobbit, and his name was Baggins. The Bagginses had lived in the neighbourhood of The Hill for time out of mind, and people considered them very respectable, not only because most of them were rich, but also because they never had any adventures or did anything unexpected: you could tell what a Baggins would say on any question without the bother of asking him. This is a story of how a Baggins had an adventure, and found himself doing and saying things altogether unexpected. He may have lost the neighbours’ respect, but he gained - well, you will see whether he gained anything in the end. The mother of our particular hobbit - what is a hobbit? I suppose hobbits need some description nowadays, since they have become rare and shy of the Big People, as they call us. They are (or were) a little people, about half our height, and smaller than the bearded Dwarves. Hobbits have no beards. There is little or no magic about them, except the ordinary everyday sort which helps them to disappear quietly and quickly when large stupid folk like you and me come blundering along, making a noise like elephants which they can hear a mile off. They are inclined to be fat in the stomach; they dress in bright colours (chiefly green and yellow); wear no shoes, because their feet grow natural leathery soles and thick warm brown hair like the stuff on their heads (which is curly)."
referencia = "En un agujero en el suelo vivía un hobbit. No un agujero húmedo, sucio, repugnante, con restos de gusanos y olor a fango, ni tampoco un agujero seco, desnudo y arenoso, sin nada en que sentarse o que comer: era un agujero-hobbit, y eso significa comodidad.Tenía una puerta redonda, perfecta como un ojo de buey, pintada de verde, con una manilla de bronce dorada y brillante, justo en el medio. La puerta se abría a un vestíbulo cilíndrico, como un túnel: un túnel muy cómodo, sin humos, con paredes revestidas de madera y suelos enlosados y alfombrados, provisto de sillas barnizadas, y montones y montones de perchas para sombreros y abrigos; el hobbit era aficionado a las visitas. El túnel se extendía serpeando, y penetraba bastante, pero no directamente, en la ladera de la colina – La Colina, como la llamaba toda la gente de muchas millas alrededor -, y muchas puertecitas redondas se abrían en él, primero a un lado y luego al otro. Nada de subir escaleras para el hobbit: dormitorios, cuartos de baño, bodegas, despensas (muchas), armarios (habitaciones enteras dedicadas a ropa), cocinas, comedores, se encontraban en la misma planta, y en verdad en el mismo pasillo. Las mejores habitaciones estaban todas a la izquierda de la puerta principal, pues eran las únicas que tenían ventanas, ventanas redondas, profundamente excavadas, que miraban al jardín y los prados de más allá, camino del río. Este hobbit era un hobbit acomodado, y se apellidaba Bolsón. Los Bolsón habían vivido en las cercanías de La Colina desde hacía muchísimo tiempo, y la gente los consideraba muy respetables, no solo porque casi todos eran ricos, sino también porque nunca tenían ninguna aventura ni hacían algo inesperado: uno podía saber lo que diría un Bolsón acerca de cualquier asunto sin necesidad de preguntárselo. Esta es la historia de cómo un Bolsón tuvo una aventura, y se encontró a sí mismo haciendo y diciendo cosas por completo inesperadas. Podría haber perdido el respeto de los vecinos, pero ganó … Bueno, ya veréis si al final ganó algo.La madre de nuestro hobbit era particular… pero, ¿qué es un hobbit? Supongo que los hobbits necesitan hoy que se los describa de algún modo, ya que se volvieron bastante raros y tímidos con la Gente Grande, como nos llaman. Son o (fueron) gente menuda de la mitad de nuestra talla, y más pequeños que los enanos barbados. Los hobbits no tienen barba. Hay poca o ninguna magia en ellos, excepto esa común y cotidiana que los ayuda a desaparecer en silencio y rápidamente, cuando gente grande y estúpida como vosotros o yo se acerca sin mirar por dónde va, con un ruido de elefantes que puede oírse a una milla de distancia. Tienden a ser gruesos de vientre; visten colores brillantes (sobre todo verde y amarillo); no usan zapatos, porque en los pies tienen unas suelas naturales de piel y un pelo espeso y tibio de color castaño, como el que les crece en las cabezas (que es rizado)."
trad_2_gt = """En un agujero en el suelo vivía un hobbit. No era un agujero repugnante, sucio y húmedo, lleno de gusanos en las puntas y con un olor a cieno, ni tampoco un agujero seco, desnudo y arenoso, sin nada en lo que sentarse o para comer: era un agujero-hobbit, y eso significa comodidad.Tenía una puerta perfectamente redonda como un ojo de buey, pintada de verde, con un pomo de latón amarillo brillante en el centro exacto. La puerta daba a un vestíbulo en forma de tubo, parecido a un túnel: un túnel muy cómodo y sin humo, con paredes revestidas y suelos de baldosas y moqueta, provisto de sillas pulidas y montones, montones de perchas para sombreros y abrigos: el hobbit estaba aficionado a los visitantes. El túnel serpenteaba más y más, yendo bastante pero no del todo recto hacia la ladera de la colina (La Colina, como la llamaba toda la gente en muchos kilómetros a la redonda) y muchas pequeñas puertas redondas se abrían fuera de él, primero a un lado y luego a otro.El hobbit no podía subir las escaleras: dormitorios, baños, sótanos, despensas (muchas), armarios (tenía habitaciones enteras dedicadas a la ropa), cocinas, comedores, todo estaba en el mismo piso y, de hecho, en el mismo pasillo.Las mejores habitaciones estaban todas en el lado izquierdo (hacia la entrada), porque eran las únicas que tenían ventanas, ventanas redondas y profundas que daban al jardín y a los prados más allá, que descendían hasta el río.Este hobbit era un hobbit muy acomodado y se llamaba Bolsón. Los Bolsón habían vivido en el barrio de The Hill desde hacía mucho tiempo, y la gente los consideraba muy respetables, no sólo porque la mayoría eran ricos, sino también porque nunca tuvieron aventuras ni hicieron nada inesperado: se podía decir qué diría Bolsón sobre cualquier pregunta sin molestarse en preguntarle. Esta es una historia de cómo Bolsón tuvo una aventura y se encontró haciendo y diciendo cosas completamente inesperadas. Puede que haya perdido el respeto de los vecinos, pero lo ganó … bueno, verás si al final ganó algo.La madre de nuestro hobbit particular - ¿qué es un hobbit? Supongo que hoy en día los hobbits necesitan alguna descripción, ya que se han vuelto raros y tímidos con la Gente Grande, como nos llaman. Son (o eran) personas pequeñas, aproximadamente la mitad de nuestra altura, y más pequeñas que los barbudos Enanos. Los hobbits no tienen barba. Hay poca o ninguna magia en ellos, excepto la del tipo cotidiano que les ayuda a desaparecer silenciosa y rápidamente cuando gente grande y estúpida como tú y yo llegamos dando tumbos, haciendo un ruido como el de los elefantes que pueden oír a una milla de distancia apagado. Tienden a tener grasa en el estómago; visten con colores brillantes (principalmente verde y amarillo), no usan zapatos, porque a sus pies les crecen suelas de cuero natural y un espeso y cálido cabello castaño como el que tienen en la cabeza (que es rizado)."""
trad_3_pt = """En un agujero en el suelo vivía un hobbit. Ni un agujero sucio, sucio, húmedo, lleno de lombrices y un olor ondulado, ni aún un agujero seco, desnudo, arenoso, sin nada en él para sentarse o para comer: era un agujero de hobbit, y eso significa comodidad.Tenía una puerta perfectamente redonda como un pórtico, pintado de verde, con un pomo de latón amarillo brillante en el centro exacto. La puerta se abría a una sala en forma de tubo como un túnel: un túnel muy cómodo sin humo, con paredes de paneles, y pisos de baldosas y alfombrado, provisto de sillas pulidas, y un montón y un montón de clavijas para sombreros y abrigos - el hobbit era aficionado a los visitantes. La herida del túnel en y en, yendo bastante pero no del todo recta en el lado de la colina - La colina, como toda la gente durante muchos kilómetros a la redonda lo llamó - y muchas pequeñas puertas redondas se abrieron fuera de ella, primero en un lado y luego en otro.No ir arriba para el hobbit: dormitorios, baños, bodegas, despensas (muchos de estos), armarios (tenía habitaciones enteras dedicadas a la ropa), cocinas, comedores, todos estaban en el mismo piso, y de hecho en el mismo pasaje.Las mejores habitaciones estaban todas a la izquierda (entrando), porque estas eran las únicas que tenían ventanas, ventanas redondas profundas mirando sobre su jardín, y prados más allá, inclinándose hacia el río.Este hobbit era un hobbit muy bueno, y su nombre era Baggins. Los Bagginses habían vivido en el barrio de The Hill durante mucho tiempo fuera de la mente, y la gente los consideraba muy respetables, no sólo porque la mayoría de ellos eran ricos, sino también porque nunca tuvieron aventuras o hicieron nada inesperado: se podía decir lo que un Baggins diría en cualquier pregunta sin la molestia de preguntarle. Esta es una historia de cómo un Baggins tuvo una aventura, y se encontró haciendo y diciendo cosas totalmente inesperadas. Puede haber perdido el respeto de los vecinos, pero ganó - bueno, usted verá si él ganó algo al final.La madre de nuestro hobbit en particular - ¿qué es un hobbit? Supongo que los hobbits necesitan alguna descripción hoy en día, ya que se han vuelto raros y tímidos de la Gente Grande, como nos llaman. Son (o eran) un poco personas, aproximadamente la mitad de nuestra altura, y más pequeños que los enanos barbudos.Los hobbits no tienen barba. Hay poca o ninguna magia en ellos, excepto el tipo cotidiano ordinario que les ayuda a desaparecer silenciosa y rápidamente cuando gente estúpida grande como tú y yo vienen haciendo un ruido como elefantes que pueden oír una milla fuera.Están inclinados a estar gordos en el estómago; visten en colores brillantes (principalmente verde y amarillo); no usan zapatos, porque sus pies crecen suelas de cuero natural y cabello castaño cálido grueso como las cosas en sus cabezas (que es rizado)."""

In [ ]:
original

'In a hole in the ground there lived a hobbit. Not a nasty, dirty, wet hole, filled with the ends of worms and an oozy smell, nor yet a dry, bare, sandy hole with nothing in it to sit down on or to eat: it was a hobbit-hole, and that means comfort. It had a perfectly round door like a porthole, painted green, with a shiny yellow brass knob in the exact middle. The door opened on to a tube-shaped hall like a tunnel: a very comfortable tunnel without smoke, with panelled walls, and floors tiled and carpeted, provided with polished chairs, and lots and lots of pegs for hats and coats - the hobbit was fond of visitors. The tunnel wound on and on, going fairly but not quite straight into the side of the hill - The Hill, as all the people for many miles round called it - and many little round doors opened out of it, first on one side and then on another. No going upstairs for the hobbit: bedrooms, bathrooms, cellars, pantries (lots of these), wardrobes (he had whole rooms devoted to clothes)

In [ ]:
##Partimos en frases los textos para poder realizar la evaluación
frasesOriginal = [x.strip() for x in original.split('.')[:-1]]
frasesOriginal

['In a hole in the ground there lived a hobbit',
 'Not a nasty, dirty, wet hole, filled with the ends of worms and an oozy smell, nor yet a dry, bare, sandy hole with nothing in it to sit down on or to eat: it was a hobbit-hole, and that means comfort',
 'It had a perfectly round door like a porthole, painted green, with a shiny yellow brass knob in the exact middle',
 'The door opened on to a tube-shaped hall like a tunnel: a very comfortable tunnel without smoke, with panelled walls, and floors tiled and carpeted, provided with polished chairs, and lots and lots of pegs for hats and coats - the hobbit was fond of visitors',
 'The tunnel wound on and on, going fairly but not quite straight into the side of the hill - The Hill, as all the people for many miles round called it - and many little round doors opened out of it, first on one side and then on another',
 'No going upstairs for the hobbit: bedrooms, bathrooms, cellars, pantries (lots of these), wardrobes (he had whole rooms dev

In [ ]:
referencia_frases = [a.strip() for a in referencia.split('.')[:-1]]
referencia_frases

['En un agujero en el suelo vivía un hobbit',
 'No un agujero húmedo, sucio, repugnante, con restos de gusanos y olor a fango, ni tampoco un agujero seco, desnudo y arenoso, sin nada en que sentarse o que comer: era un agujero-hobbit, y eso significa comodidad',
 'Tenía una puerta redonda, perfecta como un ojo de buey, pintada de verde, con una manilla de bronce dorada y brillante, justo en el medio',
 'La puerta se abría a un vestíbulo cilíndrico, como un túnel: un túnel muy cómodo, sin humos, con paredes revestidas de madera y suelos enlosados y alfombrados, provisto de sillas barnizadas, y montones y montones de perchas para sombreros y abrigos; el hobbit era aficionado a las visitas',
 'El túnel se extendía serpeando, y penetraba bastante, pero no directamente, en la ladera de la colina – La Colina, como la llamaba toda la gente de muchas millas alrededor -, y muchas puertecitas redondas se abrían en él, primero a un lado y luego al otro',
 'Nada de subir escaleras para el hobbit: 

In [ ]:
trad_2_gt = [b.strip() for b in trad_2_gt.split('.')[:-1]]
trad_2_gt

['En un agujero en el suelo vivía un hobbit',
 'No era un agujero repugnante, sucio y húmedo, lleno de gusanos en las puntas y con un olor a cieno, ni tampoco un agujero seco, desnudo y arenoso, sin nada en lo que sentarse o para comer: era un agujero-hobbit, y eso significa comodidad',
 'Tenía una puerta perfectamente redonda como un ojo de buey, pintada de verde, con un pomo de latón amarillo brillante en el centro exacto',
 'La puerta daba a un vestíbulo en forma de tubo, parecido a un túnel: un túnel muy cómodo y sin humo, con paredes revestidas y suelos de baldosas y moqueta, provisto de sillas pulidas y montones, montones de perchas para sombreros y abrigos: el hobbit estaba aficionado a los visitantes',
 'El túnel serpenteaba más y más, yendo bastante pero no del todo recto hacia la ladera de la colina (La Colina, como la llamaba toda la gente en muchos kilómetros a la redonda) y muchas pequeñas puertas redondas se abrían fuera de él, primero a un lado y luego a otro',
 'El hobb

In [ ]:
trad_3_pt = [c.strip() for c in trad_3_pt.split('.')[:-1]]
trad_3_pt

['En un agujero en el suelo vivía un hobbit',
 'Ni un agujero sucio, sucio, húmedo, lleno de lombrices y un olor ondulado, ni aún un agujero seco, desnudo, arenoso, sin nada en él para sentarse o para comer: era un agujero de hobbit, y eso significa comodidad',
 'Tenía una puerta perfectamente redonda como un pórtico, pintado de verde, con un pomo de latón amarillo brillante en el centro exacto',
 'La puerta se abría a una sala en forma de tubo como un túnel: un túnel muy cómodo sin humo, con paredes de paneles, y pisos de baldosas y alfombrado, provisto de sillas pulidas, y un montón y un montón de clavijas para sombreros y abrigos - el hobbit era aficionado a los visitantes',
 'La herida del túnel en y en, yendo bastante pero no del todo recta en el lado de la colina - La colina, como toda la gente durante muchos kilómetros a la redonda lo llamó - y muchas pequeñas puertas redondas se abrieron fuera de ella, primero en un lado y luego en otro',
 'No ir arriba para el hobbit: dormitor

#WER
A continuación vamos a medir el número mínimo de transformaciones que son necesarias para transformar la traducción generada en la de referencia.

Cuanto más bajo mejor.

In [ ]:
from evaluate import load
wer = load("wer")

In [ ]:
#Índice de Wer en la traducción automática de Google Translate
wer_score2 = wer.compute(predictions=trad_2_gt, references=referencia_frases)
wer_score2

0.44554455445544555

In [ ]:
#Índice de Wer en la traducción automática de PromptOne
wer_score3 = wer.compute(predictions=trad_3_pt, references=referencia_frases)
wer_score3

0.5188118811881188

##BLEU
Mide el solapamiento de n-gramas de la traducción de referencia y la traducción producida

In [ ]:
bleu = load("bleu")

In [ ]:
bleu_score2 = bleu.compute(predictions=trad_2_gt, references=referencia_frases)
bleu_score2

{'bleu': 0.4128030552841886,
 'precisions': [0.7208480565371025,
  0.4818181818181818,
  0.3445692883895131,
  0.25675675675675674],
 'brevity_penalty': 0.985965144294571,
 'length_ratio': 0.9860627177700348,
 'translation_length': 566,
 'reference_length': 574}

In [ ]:
bleu_score3 = bleu.compute(predictions=trad_3_pt, references=referencia_frases)
bleu_score3

{'bleu': 0.35930184755497147,
 'precisions': [0.6726296958855098,
  0.42357274401473294,
  0.2998102466793169,
  0.2172211350293542],
 'brevity_penalty': 0.9735231916601762,
 'length_ratio': 0.9738675958188153,
 'translation_length': 559,
 'reference_length': 574}

##METEOR
Mide la media armónica de precisión y recall
-	Calcula el número de unigramas coincidentes (m), el número de unigramas en traducción (Wt) y el número de unigramas en referencia (Wr)


In [ ]:
meteor = load("meteor")

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


In [ ]:
meteor_score2 = meteor.compute(predictions=trad_2_gt, references=referencia_frases)
meteor_score2

{'meteor': 0.710133585241393}

In [ ]:
meteor_score3 = meteor.compute(predictions=trad_3_pt, references=referencia_frases)
meteor_score3

{'meteor': 0.6592587307449456}

## BERT Score
Queremos calular la similitud entre traducciones.
BERTScore: calcula la similitud entre embeddings de Bert (Cuanto mayor sea la similitud entre la traducción generada y la traducción de referencia significa que la traducción es buena).

In [ ]:
bertscore = load("bertscore")

In [ ]:
bert_score2 = bertscore.compute(predictions=trad_2_gt, references=referencia_frases,lang="es")
bert_score2

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/996k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/714M [00:00<?, ?B/s]

{'precision': [1.0,
  0.9182800054550171,
  0.90938800573349,
  0.8896709680557251,
  0.8762730360031128,
  0.9185417890548706,
  0.8948869705200195,
  0.9463254809379578,
  0.9089497327804565,
  0.9534065127372742,
  0.9091652631759644,
  0.9186438918113708,
  0.8291003704071045,
  1.0000001192092896,
  0.8785036206245422,
  0.8804786801338196],
 'recall': [1.0,
  0.9400411248207092,
  0.90960693359375,
  0.8778767585754395,
  0.881233811378479,
  0.9201735258102417,
  0.8942732810974121,
  0.929012656211853,
  0.8839410543441772,
  0.9041553139686584,
  0.9099560379981995,
  0.8864481449127197,
  0.8335188627243042,
  1.0000001192092896,
  0.8737958073616028,
  0.8792362213134766],
 'f1': [1.0,
  0.9290331602096558,
  0.9094974398612976,
  0.8837345242500305,
  0.8787463903427124,
  0.9193569421768188,
  0.8945800065994263,
  0.9375892281532288,
  0.8962709903717041,
  0.9281280040740967,
  0.9095604419708252,
  0.9022588729858398,
  0.8313037157058716,
  1.0000001192092896,
  0.8761

In [ ]:
bert_score3 = bertscore.compute(predictions=trad_3_pt, references=referencia_frases,lang="es")
bert_score3

{'precision': [1.0,
  0.9096742868423462,
  0.8652496933937073,
  0.882415235042572,
  0.8248111605644226,
  0.9239621758460999,
  0.8764472007751465,
  0.8680148720741272,
  0.8780391216278076,
  0.9527713656425476,
  0.858864426612854,
  0.9130902290344238,
  0.8466671109199524,
  1.0000001192092896,
  0.8741633892059326,
  0.8692276477813721],
 'recall': [1.0,
  0.9017808437347412,
  0.8654589056968689,
  0.868232011795044,
  0.8134229779243469,
  0.921212911605835,
  0.8677095770835876,
  0.835304319858551,
  0.8691019415855408,
  0.9098694920539856,
  0.8697825074195862,
  0.8845689296722412,
  0.8534232378005981,
  1.0000001192092896,
  0.8364077806472778,
  0.8688263893127441],
 'f1': [1.0,
  0.9057103991508484,
  0.8653542995452881,
  0.8752661347389221,
  0.8190774917602539,
  0.9225855469703674,
  0.8720564842224121,
  0.8513455390930176,
  0.8735476732254028,
  0.9308263659477234,
  0.8642889261245728,
  0.8986033797264099,
  0.8500317335128784,
  1.0000001192092896,
  0.854

In [ ]:
import numpy as np
np.mean(bert_score2['f1']),np.std(bert_score2['f1'])

(0.9110037684440613, 0.04192610443751822)

In [ ]:
np.mean(bert_score3['f1']),np.std(bert_score3['f1'])

(0.8907868750393391, 0.04944280839669548)

## BLEURT

In [ ]:
bleurtscore = load("bleurt")

In [ ]:
bleurt_score2 = bleurtscore.compute(predictions=trad_2_gt, references=referencia_frases)
bleurt_score2

{'scores': [1.0417605638504028,
  0.14160360395908356,
  0.18586009740829468,
  0.1913139522075653,
  -0.0608946718275547,
  0.3101072311401367,
  0.19887709617614746,
  0.5905261635780334,
  0.26991623640060425,
  0.46724265813827515,
  0.39440762996673584,
  0.218680739402771,
  -0.17733970284461975,
  1.0333631038665771,
  0.1779910922050476,
  0.31665802001953125]}

In [ ]:
bleurt_score3 = bleurtscore.compute(predictions=trad_3_pt, references=referencia_frases)
bleurt_score3

{'scores': [1.0417605638504028,
  0.2632640600204468,
  0.060560192912817,
  0.20052748918533325,
  -0.16013219952583313,
  0.35697275400161743,
  0.1373942792415619,
  0.12289384007453918,
  0.012723442167043686,
  0.3536975383758545,
  0.24433612823486328,
  0.2230350524187088,
  -0.20638784766197205,
  1.0333631038665771,
  0.20755073428153992,
  0.08357682824134827]}

###Calculamos la media y la desviación típica de los valores y modelos utilizados

In [ ]:
np.mean(bleurt_score2['scores']),np.std(bleurt_score2['scores'])

(0.3312546133529395, 0.3196410668824179)

In [ ]:
np.mean(bleurt_score3['scores']),np.std(bleurt_score3['scores'])

(0.24844599748030305, 0.3347670306245857)

##Resumen de los resultados
En la siguiente tabla vemos un resumen de los resultados:

||Google Translate|PromptOne|
|---|---|---|
| WER | 0.44 | 0.51|
| BLEU | 0.41 | 0.35|
| METEOR |0.71|0.65|
| BERT Score | 0.911 (0.041) | 0.89 (0.049)|
| BLEURT | 0.33 (0.31) | 0.24 (0.33)|

En este caso podemos decir que:
- WER: el número es menor en la evaluación de la traducción de Google Translate por lo que es mejor traducción según este método. (Encuentra menos errores)
- BLEU: el número es mayor en la evaluación de la traducción de Google Translate por lo que es mejor traducción según este método. (Hay mayor solapamiento en el modelo de Google Translate).
- METEOR: dado que el número es mayor en el primer modelo, la traducción de Google Translate es mejor que la de PromptOne (una vez más).
- BERTScore: el primer modelo tiene mayor similitud entre la traducción generada y la de referencia, por lo que es mejor Google Translate.
- BLEURT:el cálculo de la media del segundo modelo (PromptOne) es ligeramente peor que el de Google Translate. Esta es mejor traducción.